elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,247 @@
+            return visitPredicate((FilterPredicate) normalized);
+        }
+    }
+
+    /**
+     * Returns whether or not a pair of {@link FilterExpression}s is a pure {@code HAVING} expression, i.e. no
+     * {@code WHERE} clause.
+     *
+     * @param filterExpression  The pair to check
+     *
+     * @return {@code true} if there is {@code HAVING} expression only and not {@code WHERE} expression.
+     */
+    private static boolean isHavingExpression(Pair<FilterExpression, FilterExpression> filterExpression) {
+        return filterExpression.getLeft() == null && filterExpression.getRight() != null;
+    }
+
+    /**
+     * Returns whether or not a pair of {@link FilterExpression}s is a pure {@code WHERE} expression, i.e. no
+     * {@code HAVING} clause.
+     *
+     * @param filterExpression  The pair to check
+     *
+     * @return {@code true} if there is {@code WHERE} expression only and not {@code HAVING} expression.
+     */
+    private static boolean isWhereExpression(Pair<FilterExpression, FilterExpression> filterExpression) {
+        return filterExpression.getLeft() != null && filterExpression.getRight() == null;
+    }
+
+    /**
+     * Returns the {@code WHERE} expression from the pair of {@link FilterExpression}s.
+     * <p>
+     * See class documentation for more details on the "pair".
+     *
+     * @param filterExpressionPair  The pair to extract the {@code WHERE} expression from
+     *
+     * @return the {@code WHERE} constraint
+     */
+    private static FilterExpression extractWhereExpression(Pair<FilterExpression, FilterExpression> filterExpressionPair) {
+        return filterExpressionPair.getLeft();
+    }
+
+    /**
+     * Returns the {@code HAVING} expression from the pair of {@link FilterExpression}s.
+     * <p>
+     * See class documentation for more details on the "pair".
+     *
+     * @param filterExpressionPair  The pair to extract the {@code HAVING} expression from
+     *
+     * @return the {@code HAVING} constraint
+     */
+    private static FilterExpression extractHavingExpression(Pair<FilterExpression, FilterExpression> filterExpressionPair) {
+        return filterExpressionPair.getRight();
+    }
+
+    /**
+     * Returns whether or not a {@link FilterPredicate} corresponds to a {@code HAVING} clause in JPQL query.
+     * <p>
+     * A {@link FilterPredicate} corresponds to a {@code HAVING} clause iff the predicate field has
+     * {@link MetricAggregation} annotation on it.
+     *
+     * @param filterPredicate  The terminal filter expression to check for
+     *
+     * @return {@code true} if the {@link FilterPredicate} is a HAVING clause
+     */
+    private boolean isHavingPredicate(final FilterPredicate filterPredicate) {
+        String entityName = filterPredicate.getEntityType().getName();
+        String fieldName = filterPredicate.getField();
+
+        try {
+            return getEntityDictionary().getFieldAnnotations(entityName, fieldName).stream()
```

Comment:
There is a method in EntityDictionary called `getAttributeOrRelationAnnotation` that you can use here instead.